### ISSUE TITLE
cvxopt-only case for CI testing 

### ISSUE THREAD

**Issue by whoburg** [convexopt/gpkit#312]
Heard from a new windows user that when only cvxopt is installed, we get a unit test failure related to mosek.

 - update continuous integration to include a machine with only cvxopt installed (?) @galbramc 
 - find what's causing this issue and fix it (@whoburg knows how, but waiting so we can reproduce this failure in CI)

**Comment by whoburg** [comment]
![gpkit_tests_error](https://cloud.githubusercontent.com/assets/6042395/9064142/292eab7e-3a99-11e5-9f0f-75a31b127aac.png)

**Comment by galbramc** [comment]
Would you also want to have testing that only tests mosek without cvxopt? This is what is actually happening on the windows test machine because I could not get cvxopt compiled on windows.

How did the user get cvxopt installed?!

**Comment by whoburg** [comment]
I believe they used Pythonxy (https://code.google.com/p/pythonxy/), which includes an option to install cvxopt as part of the pythonxy install.

Testing with cvxopt (and no mosek) is the most pressing special case because cvxopt is open source, making it the only free option for non-academic users.

**Comment by galbramc** [comment]
I currently have anaconda installed on the Windows machine. Do you suggest
I uninstall that and install pythonxy instead? Pythonxy says it does not
play nice with other python installs.

**Comment by whoburg** [comment]
No, I think the current CI setup is great and we should stick with it. If we did pythonxy tests I would think they should be in addition. Does a virtualenv setup make sense here, or does that defeat the purpose of on-hardware CI?

**Comment by galbramc** [comment]
Yeah we could setup virtualenv. I could probably use some help though from
someone, I have never used virtualenv, particularly on a windows machine. I
presume we would need to install pythonxy in the virtualenv every time we
run the test?

**Comment by whoburg** [comment]
This might go completely against the spirit of continuous integration, but for this add-on unit testing case, I think it might be fine to just run the unit tests (on the same installation every time), to make sure we haven't introduced code that assumes mosek exists.

Does that seem reasonable? And more importantly, would it make this a lot easier?

**Comment by galbramc** [comment]
I added another axis so we can try to test cvxopt by it self, and
cvxopt+mosek. I don't have an explicit mosek only test on linux and OSX,
but since I could not get cvxopt installed on the windows machine it is
only testing with mosek.

I can't easily test things at the moment until someone creates a pull
request and triggers the builds. This should of course fail unless the
mosek dependency in the tests is fixed.

**Comment by whoburg** [comment]
Cool, so just to confirm, the "cvxopt" axis is only cvxopt, and "mosek" is cvxopt+mosek?

**Comment by galbramc** [comment]
Yeah, I didn't feel like making the combo explicit as the name should be
really long.

How do I expose the system numpy to the virtialenv? I also have cvxopt
installed with --user which I'm guessing the virtualenv is also going to
ignore.

**Comment by whoburg** [comment]
is it an anaconda virtualenv? If so, just give the packages, e.g.

`conda create -n cvxopt_only python2.7 numpy pint`

then you would

`source activate cvxopt_only`

and install cvxopt (`python setup.py install`) into the activated environment

Hope this is helpful (?) .

**Comment by galbramc** [comment]
Not using Anaconda on the Linux or the Mac.

**Comment by whoburg** [comment]
With standard virtualenv you should be able to simply pip install:

```
source <env name>/bin/activate
pip install numpy
pip install pint
cd <path to cvxopt>
python setup.py install
```

**Comment by whoburg** [comment]
@galbramc we can debug in person if easier

**Comment by galbramc** [comment]
Yes let's do that. I have a meeting at 3, I can let you know when I'm back.

**Comment by galbramc** [comment]
I'm back.

**Comment by galbramc** [comment]
I added the flag --system-site-packages to virtualenv. Hopefully this solves the problem. Give it a try.

**Comment by galbramc** [comment]
So I don't know how to get build.py to ignore mosek. Right now it just looks in the home directory to see if it's there, and if so gpkit assumes that mosek is installed and tries to use it. This means I would have to modify the testing script to delicately rename the mosk directory in the home directory to avoid gpkit configuring it self to use moske. The mucking with .bashrc also precludes me from preventing gpkit from finding mosek.

**Comment by galbramc** [comment]
Do all these tests pass on your compter (https://acdl.mit.edu/csi/job/gpkit/buildnode=reynolds,optimizer=mosek/224/console). Looking at the output it does not look like these are errors from not finding mosek (this is the case that should be testing mosek).

**Comment by galbramc** [comment]
Can we add a flag to build.py to tell it to not look for mosek?

**Comment by whoburg** [comment]
@galbramc, yes, all those tests pass on my computer. That's the first time I've seen those failures on jenkins -- looks like solver issues to me, though the errors are admittedly difficult to parse/understand.

@bqpd what do you think of adding a `--nomosek` flag to build.py and/or setup.py to deal with this issue? Seems reasonable to me but I may be missing a problem.

**Comment by galbramc** [comment]
Adding the flag --nomosek to setup.py should do the trick.

I don't know why the tests with mosek failed... Could we try one more time? I tested things out manually on the machine and it worked. I made a couple of small changes to the jenkins script.

**Comment by whoburg** [comment]
hmm, still seeing similar errors

**Comment by galbramc** [comment]
I removed the virtualenv. Can we give it a try again?

**Comment by galbramc** [comment]
I was able to reproduce the error. The problem is I am running multiple instances of gpkit at the same time. It appears you are constantly creating and deleting the folder ~/gpkit_tmp. This fundamentally means I cannot have multiple tests running at the same time on the same computer. So I can either serialize the testing, or you can use some other folder other than ~/gpkit_tmp as a temporary folder. I believe python provides an interface to request a temporary folder in /tmp with a unique name, but don't recall for sure how to use it.

**Comment by galbramc** [comment]
I've removed the duplicate testing until we get the --nomosek flag. The test should work again, but we are back to testing cvxopt+mosek at the moment.

**Comment by galbramc** [comment]
Sorry, forgot to fix one line in the script. Should work now.

**Comment by galbramc** [comment]
One more try. I had forgotten to uninstall mosek on the mac and I think that was messing with the virtualenv

**Comment by galbramc** [comment]
Try again.

**Comment by whoburg** [comment]
Thanks @galbramc for getting CI back to a working state. Here's an updated list of todos. Waiting on @bqpd's thoughts on adding `--nomosek` flag to setup.py.

 - [ ] @bqpd what do you think of adding a `--nomosek` flag to build.py and/or setup.py to deal with this issue? Seems reasonable to me but I may be missing a problem.
 - [ ] update continuous integration to include a machine with only cvxopt installed (?) @galbramc 
 - [ ] original error was in `t_gp_solution_array.TestGPSolutionArray.test_table` due to `solver='mosek'`. @whoburg fixed this in e5acc459b496a7025afe007f295a6650c7825d22. Temporarily add back `solver='mosek'` on a branch to confirm CI catches the error (but don't merge that change).
 - [ ] It appears we are constantly creating and deleting the folder ~/gpkit_tmp. This fundamentally means we cannot have multiple tests running at the same time on the same computer. So we can either serialize the testing, or use some other folder other than ~/gpkit_tmp as a temporary folder. I believe python provides an interface to request a temporary folder in /tmp with a unique name, but don't recall for sure how to use it.

**Comment by bqpd** [comment]
The `--nomosek` flag would only be used for setting up test environments? Most users won't run `build.py` themselves.

**Comment by galbramc** [comment]
Yes. I only need --nomosek to setup testing workout it. I would need it for
setup.py. It's either that or I serialize thinks and very carefully rename
~/mosek to prevent setup from finding it. I really do not like that
solution.

**Comment by whoburg** [comment]
@bqpd, is your point that we know what's causing the original unit test failure (indeed, we do), and so we might as well just fix it and not have cvxopt-only testing be part of our CI?

That will work short term but don't you think there's value in CI testing a cvxopt only setup?

**Comment by galbramc** [comment]
If you want gpkit to always work with only cvxopt, then you want to make
sure you test with only cvxopt to make sure a similar bug to what you just
fixed does not get reintroduced.

I don't run build.py when I setup the testing. I use setup.py to mimic what
a user would do. I just need the --nomosek flag on setup.py so I can test
the situation where a user does not have mosek on their system, but I don't
want to constantly rename the ~/mosek folder to prevent setup.py from
finding it.

**Comment by bqpd** [comment]
Naw, I meant changing the unit test so that it doesn't unnecessarily call on MOSEK, as below.

We can definitely add the `--nomosek` flag for the case where mosek is installed but we don't want to access it.

```python
    def test_table(self):
        x = Variable('x')
        gp = Model(x, [x >= 12])
        # sol = gp.solve(solver='mosek', verbosity=0)
        sol = gp.solve(verbosity=0)
        tab = sol.table()
        self.assertTrue(isinstance(tab, str))
```

**Comment by galbramc** [comment]
I think it's important to test gpkit with only cvxopt since that is a supported mode. If that fix to the unit tests works then thats fine with me, I don't know the details of gpkit.

I can use a virtualevn where mosek is not installed for testing without mosek. But I will have ~/mosek. If you can look at the python environment to decide if testing should be done with mosek or not works for me. Otherwise I need the explicit flag --nomoske to tell gpkit to not use mosek.

I don't know how you choose the optimizers. Does the unit testing test both cvxopt and mosek when both are available?

**Comment by bqpd** [comment]
Yup, it does.

**Comment by bqpd** [comment]
I'll add the `--nomosek` flag, then! Alternately, if you're running a bash script to set this all up, you could edit the `gpkit/env/settings` file to change which solvers are used.

**Comment by galbramc** [comment]
So if I just remove mosek from gpkit/env/settings you won't test with mosek?

**Comment by galbramc** [comment]
Let me try adding the echo command. I can easily do that.

**Comment by bqpd** [comment]
Yup, it only tests solvers that are in that file.

**Comment by galbramc** [comment]
So I tested things on the master branch. If I don't have mosek in the DYLD_LIBRARY_PATH on the mac I get these errors:

======================================================================
ERROR [0.002s]: test_table (gpkit.tests.t_gp_solution_array.TestGPSolutionArray)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jenkins/workspace/gpkit/buildnode/macys/optimizer/cvxopt/gpkit/tests/t_gp_solution_array.py", line 70, in test_table
    sol = gp.solve(printing=False)
  File "/Users/jenkins/workspace/gpkit/buildnode/macys/optimizer/cvxopt/gpkit/geometric_program.py", line 353, in solve
    return self._solve(*args, **kwargs)
  File "/Users/jenkins/workspace/gpkit/buildnode/macys/optimizer/cvxopt/gpkit/geometric_program.py", line 379, in _solve
    from ._mosek import expopt
  File "/Users/jenkins/workspace/gpkit/buildnode/macys/optimizer/cvxopt/gpkit/_mosek/expopt.py", line 26, in <module>
    raise ImportError("Could not load MOSEK library: "+repr(e))
ImportError: Could not load MOSEK library: ImportError(OSError('dlopen(/Users/jenkins/.gpkit/expopt.so, 10): Library not loaded: @loader_path/libmosek64.7.1.dylib\n  Referenced from: /Users/jenkins/.gpkit/expopt.so\n  Reason: image not found',),)

Oddly enough it seems that it working on the Linux machine. 

Also, I still have to run things serially due to the gpkit_tmp folder. This means nobody can ever run gpkit in parallel on a computer to solve multiple problems simultaneously.

**Comment by galbramc** [comment]
@whoburg I used @bqpd trick to only use cvxopt with the command

echo "installed_solvers : cvxopt" > $GPKIT/env/setting 

However, the gpkit master branch still seems to try to test mosek, or at least it tries to load the shared library even if it isn't getting used. I'm only getting errors on OSX on the moment though. You can see the output here:

https://acdl.mit.edu/csi/job/gpkit/buildnode=macys,optimizer=cvxopt/236/console

**Comment by whoburg** [comment]
Is there a typo in setting vs settings in that command?

should
`echo "installed_solvers : cvxopt" > $GPKIT/env/setting`
instead be
`echo "installed_solvers : cvxopt" > $GPKIT/env/settings`
?

@bqpd I already removed the mosek call from that unit test in https://github.com/convexopt/gpkit/commit/e5acc459b496a7025afe007f295a6650c7825d22

**Comment by bqpd** [comment]
Yup, it should be `settings`, my mistake!

**Comment by galbramc** [comment]
Looks like its working now after the typo got fixed. I believe you have testing of cvxopt alone now.

The testing is still serial due to gpkit_tmp/. Maybe that should be opened as another issue.

**Comment by whoburg** [comment]
Working perfectly as confirmed by #314! 

Remaining checkboxes are a) the --nomosek flag (@galbramc seems like we can skip this for now?), and b) the gpkit_tmp serial naming issue.

**Comment by bqpd** [comment]
gpkit_tmp is fixed in the GP refactor, which is ready for code review and
will hopefully be merged soon.

**Comment by whoburg** [comment]
Great! Dropping --nomosek for now.

**Comment by bqpd** [comment]
What still needs to be done in this issue?

**Comment by galbramc** [comment]
I believe that this issue has been resolved, and @whoburg has opened a new issue in regards to parallel testing.

**Comment by bqpd** [comment]
Great!

**Comment by whoburg** [comment]
@bqpd confirm the third checkbox above (re: gpkit_tmp) should be checked off?

**Comment by bqpd** [comment]
Yup, we create proper tmp folders now
